### DISCUSSION TITLE
Pyright complains about typing of middleware argument in constructor

### DISCUSSION THREAD

**Discussion by martimors** [encode/starlette#2451] [Potential Issue] [answered]
Error reported independently by both `pyright` and the pycharm type-checker, I assume it is also caught by `mypy`.

Repro example:

```py
class AccessLogMiddleware(BaseHTTPMiddleware):
    async def dispatch(
        self, request: Request, call_next: Callable[[Request], Awaitable[Response]]
    ) -> Response:
        # Does nothing
        response = await call_next(request)
        return response

app = Starlette(middleware=[Middleware(AccessLogMiddleware)]
```

Errors reported by pyright:

```
error: Argument of type "type[AccessLogMiddleware]" cannot be assigned to parameter "cls" of type "type[_MiddlewareClass[P@__init__]]" in function "__init__"
error: Param spec "P@__init__" has no bound value
```

**Comment by Kludex** [comment]
I can't reproduce it. Are you in the latest pyright version?

**Reply by koldakov** [reply]
Looks like I'm facing the same issue, even without any extra instruments.

The problem is that in the commit 866a15f79c7f15b5dba695ed2180922b7191a19e there was a change for type hinting from

```python
middleware_class: type,
```

to

```python
middleware_class: typing.Type[_MiddlewareClass[P]],
```

So it waits for middlewares to be inherited from _MiddlewareClass, what is wrong I believe.

From the docs:

```python
from starlette.applications import Starlette
from starlette.middleware.base import BaseHTTPMiddleware


class CustomMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
        response = await call_next(request)
        response.headers['Custom-Header'] = 'Example'
        return response


app = Starlette()
app.add_middleware(CustomMiddleware)
```

Gives:

```
Expected type 'Type[_MiddlewareClass]' (matched generic type 'Type[_MiddlewareClass[ParamSpec("P")]]'), got 'Type[CustomMiddleware]' instead
```

Environment:
```
python==3.12.0
starlette==0.36.3
```

**Reply by btakita** [reply]
> Looks like I'm facing the same issue, even without any extra instruments.

I also see this issue using a Jetbrains editor, such as Idea or Pycharm.

```
Expected type 'Type[_MiddlewareClass[ParamSpec("P")]]', got 'Type[CORSMiddleware]' instead 
```

Note that the middleware is `starlette.cors.CORSMiddleware`.

**Reply by Kludex** [reply]
This should be fixed now... Is it? I can't reproduce it.

**Reply by just-maiyak** [reply]
This seems to be fixed in [`commit 427a8d`](https://github.com/encode/starlette/commit/427a8dcf357597df27b2509b1ac436caf7708300).

**Reply by maj-github** [reply]
> This seems to be fixed in [`commit 427a8d`](https://github.com/encode/starlette/commit/427a8dcf357597df27b2509b1ac436caf7708300).

Hey! We are on 0.46.1 right now, and we still have the same issue. `Expected type '_MiddlewareFactory[ParamSpec("P")]', got 'Type[RequestMiddleware]' instead`. `RequestMiddleware` is our middleware class, a subclass of `BaseHTTPMiddleware`.

**Comment by just-maiyak** [comment]
I have a similar error popping up, I've made an Issue over on [asgi-loggers' repo](https://github.com/Kludex/asgi-logger/issues/45) documenting my case. Since I posted, I figured out it started since [commit 866a15](https://github.com/encode/starlette/commit/866a15f79c7f15b5dba695ed2180922b7191a19e).

**Comment by AdmWinther** [comment]
I am facing the same issue in FastAPI. I have made this middleware

```
from starlette.middleware.base import BaseHTTPMiddleware
class MyCORSMiddleware(BaseHTTPMiddleware):
    async def dispatch(self, request, call_next):
```

and when I use it like this:
```
app.add_middleware(MyCORSMiddleware)
```
I get the warning related to type mismatch.

**Reply by M-Studley** [reply]
Any solutions?  I hate doing #noqa in my code

**Reply by koldakov** [reply]
> Any solutions? I hate doing #noqa in my code

For now now, I described why this happens [here](https://github.com/encode/starlette/discussions/2451#discussioncomment-8400688)

Or you can use cast, but it's better to use noqa )

**Reply by NookOps** [reply]
Anybody found a solution for this?

**Reply by Kludex** [reply]
I have no issue with mypy 1.14.1. What's the problem here?

**Reply by NookOps** [reply]
I'm getting the error reported there:

[https://github.com/encode/starlette/discussions/2451#discussioncomment-8316875]()

**Reply by wisess** [reply]
I have same problem 
Expected type '_MiddlewareFactory[**P]', got 'type[RequestMiddleware]' instead
Any ideas?

**Comment by murthy-git** [comment]
It is not an error, some kind of warning. My app just works fine with this warning.

**Comment by Kludex** [comment, marked as answer]
Is this fixed now?

**Reply by koldakov** [reply]
All good now, thank you 😄

**Comment by strokirk** [comment]
Just as a side note: I've started seeing it with the recent version of `ty`, `ty 0.0.1-alpha.26 (b225fd8b4 2025-11-10)`. 

It's likely due to their incomplete support for using `type` and `Protocol` together: https://github.com/astral-sh/ty/issues/903

```
error[invalid-argument-type]: Argument to bound method `add_middleware` is incorrect
   --> app/main.py:159:9
    |
158 |     app.add_middleware(
159 |         CORSMiddleware,
    |         ^^^^^^^^^^^^^^ Expected `_MiddlewareFactory[Unknown]`, found `<class 'CORSMiddleware'>`
160 |         allow_origins=["*"],
161 |         allow_credentials=True,
    |
info: Method defined here
   --> .venv/lib/python3.13/site-packages/starlette/applications.py:123:9
    |
121 |         self.router.host(host, app=app, name=name)  # pragma: no cover
122 |
123 |     def add_middleware(
    |         ^^^^^^^^^^^^^^
124 |         self,
125 |         middleware_class: _MiddlewareFactory[P],
    |         --------------------------------------- Parameter declared here
126 |         *args: P.args,
127 |         **kwargs: P.kwargs,
    |
info: rule `invalid-argument-type` is enabled by default
```
